### PR TITLE
String#rpartition

### PR DESCRIPTION
**Pull request by johnjansen**

### CODE REVIEW

File: src/string.cr [hunk 1]
```diff
@@ -2200,6 +2200,52 @@ class String
...
+    pos = self.rindex(search) || 0
+    search_size = search.is_a?(Char) ? 1 : search.size
+
+    if pos == 0 \
+      || pos == self.size-1 \
+      || pos + search_size > self.size - 1
+      {"", "", self}
+    else
+	    pre = self[0 .. (pos - 1)]
+  	  post = self[(pos + search_size) .. -1]
```
Comment (Sija) [inline src/string.cr]: Please run the formatter tool.

File: spec/std/string_spec.cr [hunk 2]
```diff
@@ -627,6 +627,33 @@ describe "String" do
     end
   end
 
+  describe "rpartition" do
+    describe "by char" do
+      "hello".rpartition('l').should eq ({"hel", "l", "o"})
+      "hello".rpartition('o').should eq ({"", "", "hello"})
+      "hello".rpartition('h').should eq ({"", "", "hello"})
```
Comment (bcardiff) [inline spec/std/string_spec.cr]: Why if the char is at the beginning or at the end of the string is the same as if they were not found at all?

```ruby
> "hello".rpartition("o")
=> ["hell", "o", ""]
> "hello".rpartition("h")
=> ["", "h", "ello"]
```

File: src/string.cr [hunk 3]
```diff
@@ -2200,6 +2200,52 @@ class String
...
+  # "hello".rpartition("l")         #=> ["hel", "l", "o"]
+  # "hello".rpartition("x")         #=> ["", "", "hello"]
+  # "hello".rpartition(/.l/)        #=> ["he", "ll", "o"]
+  # ```
+  def rpartition(search : (Char | String)) : Tuple(String, String, String)
+    pos = self.rindex(search) || 0
+    search_size = search.is_a?(Char) ? 1 : search.size
+
+    if pos == 0 \
+      || pos == self.size-1 \
```
Comment (bcardiff) [inline src/string.cr]: there are some inconsistencies in the formatting. Use the crystal formatter tool otherwise the CI will complain.

Comment (johnjansen) [inline src/string.cr]: thanks Brian, i wasn't aware of the formatting tool i'm afraid, so apologies, i will look into that asap

File: spec/std/string_spec.cr (hunk 2 above)
Comment (johnjansen) [inline spec/std/string_spec.cr]: thats me overthinking ... shoulda checked ruby ;-)

File: src/string.cr [hunk 4]
```diff
@@ -2205,45 +2205,61 @@ class String
...
-    if pos == 0 \
-      || pos == self.size-1 \
-      || pos + search_size > self.size - 1
-      {"", "", self}
-    else
-	    pre = self[0 .. (pos - 1)]
-  	  post = self[(pos + search_size) .. -1]
+    pre = ""
+    mid = ""
+    post = ""
```
Comment (Sija) [inline src/string.cr]: `pre, mid, post = {"", "", ""}`

File: src/string.cr [hunk 5]
```diff
@@ -2200,6 +2200,68 @@ class String
...
+  # "hello".rpartition("x")  # => ["", "", "hello"]
+  # "hello".rpartition(/.l/) # => ["he", "ll", "o"]
+  # ```
+  def rpartition(search : (Char | String)) : Tuple(String, String, String)
+    pos = self.rindex(search)
+    search_size = search.is_a?(Char) ? 1 : search.size
+
+    pre = ""
+    mid = ""
+    post = ""
```
Comment (Sija) [inline src/string.cr]: `pre, mid, post = {"", "", ""}`

File: src/string.cr [hunk 6]
```diff
@@ -2200,6 +2200,68 @@ class String
...
+    if pos.nil?
+      post = self
+    elsif pos == 0
+      mid = search.to_s
+      post = self[(pos + search_size)..-1]
+    else
+      pre = self[0..(pos - 1)]
+      mid = search.to_s
+      post = self[(pos + search_size)..-1]
+    end
```
Comment (Sija) [inline src/string.cr]: IMO would look better with `case` switch:
```cr
case pos
when .nil?
  # ...
when 0
  # ...
else
  # ...
end
```

File: src/string.cr [hunk 7]
```diff
@@ -2200,6 +2200,66 @@ class String
     match_result.try &.begin(0)
   end
 
+  # Searches sep or pattern (regexp) in the string from the end of the string,
+  # and returns the part before it, the match, and the part after it.
+  # If it is not found, returns two empty strings and str.
+  #
+  # ```
+  # "hello".rpartition("l")  # => ["hel", "l", "o"]
```
Comment (luislavena) [inline src/string.cr]: The example shows that `rpartition` returns an array, but that is not the case. Please check that the code samples in the docs matches the real output.

File: spec/std/string_spec.cr [hunk 8]
```diff
@@ -627,6 +627,31 @@ describe "String" do
...
+      "hello".rpartition("l").should eq ({"hel", "l", "o"})
+      "hello".rpartition("x").should eq ({"", "", "hello"})
+      "hello".rpartition("o").should eq ({"hell", "o", ""})
+      "hello".rpartition("h").should eq ({"", "h", "ello"})
+      "hello".rpartition("ll").should eq ({"he", "ll", "o"})
+      "hello".rpartition("lo").should eq ({"hel", "lo", ""})
+      "hello".rpartition("he").should eq ({"", "he", "llo"})
+    end
+
+    describe "by regex" do
```
Comment (bcardiff) [inline spec/std/string_spec.cr]: One more corner case to ensure the regex is matched from right to left would be:
```crystal
"hello".rpartition(/l./).should eq ({"hel", "lo", ""})
```

All the existing test cases work fine if the code would make a lookup from pos=0 to length-1.

Comment (johnjansen) [inline spec/std/string_spec.cr]: hi @bcardiff, I was assuming that there was a vague probability that the split point would more likely be toward the right hand end of the string and this would save unnecessary iterations. what are your feelings on this?

Comment (bcardiff) [inline spec/std/string_spec.cr]: I am ok with the implementation BTW.

Comment (johnjansen) [inline spec/std/string_spec.cr]: ok great, we have the new test, and we are all green :-)

File: src/string.cr [hunk 9]
```diff
@@ -2200,6 +2200,66 @@ class String
...
+
+    while pos >= 0
+      self[pos..-1].scan(search) do |m|
+        match_result = m
+      end
+      match_result
+      break unless match_result.nil?
+      pos -= 1
+    end
+    match_result
```
Comment (Sija) [inline src/string.cr]: is this line necessary?

Comment (johnjansen) [inline src/string.cr]: good catch, debugging in playground ... thats going away right now

**Comment by bcardiff** [comment]
@johnjansen as soon as specs pass I am ok with this. Note that the history shall be squashed. I could handle via github directly, unless, you prefer to do that on your own.

**Comment by johnjansen** [comment]
@bcardiff ill let you do the merge, and happy to see all the attempts squashed into oblivion ;-)

**Comment by johnjansen** [comment]
@bcardiff is it just a matter of clicking "close and comment" to merge the PR ?

**Comment by asterite** [comment]
@johnjansen Thank you for this!

A question I had for some time now: why two empty strings are returned when the match is not found? Can we maybe return `nil` in this case? What's the use case for retuning two empty strings? I ask because I'm pretty sure there must be a reason.

**Comment by johnjansen** [comment]
Hi @asterite 

I wondered the same, but fell back to the same behavior as ruby ... much as i'm enthusiastic to get my first contribution merged, this probably warrants more discussion. I cant formulate a good argument either way without contradicting myself in the process :-)

J

**Comment by bcardiff** [comment]
**TL;DR 1: I think is ok to return a non-nilable tuple**
 
I am not sure if rpartition needs to return nil. If we try to use it for cycling, it sounds like returning nil seems reasonable, to stop iteration. But due to the nature of unpacking the result using it in a loop the code looks weird:

```crystal
s = "abc.def.ghi.jkl"

# if rpartition returned nil
while m = s.rpartition "."
  s, _, suffix = m
  suffix # => "jkl" , "ghi" , "def"
end

s # => "abc"
```

I would expect a block alternative to be used to cycle the results. `#each_rpartition(Char|String, &block)` or something like that that will yield all the partitions.

And last, for keeping the non-nillable tuple, a partition always exists. if you concat the 3 parts you get the original string. If the method was named "match" or something like that I would agree that we need a `nil` to represent the no-matching result.

**TL;DR 2: I think #partition and #rpartition should be equal when there is no match**

Maybe if no string is matched the result could be `{self, "", ""}` so the `rpartition` is used to split the right part **after** the separator. Like extracting the extension of a file, when no extension is found, getting an empty string seems appropriate. Also that would mean that `#partition` and `#rpartition` return the same result when there is no match at all (i would like that). But all this would defer from ruby, I am sure they had some reason for this. NB: currently there is no `String#partition`.

**Comment by johnjansen** [comment]
I'll write `partition` too, but it will be later in the week ... if you want to ticket it and assign it to me :-)

**Comment by johnjansen** [comment]
adding a block version of both could be a third task and we could agree on the nilability of elements ... once again, happy to pick that up too

**Comment by bcardiff** [comment]
@johnjansen  👍  on `#partition` for later.  But I would say to no worries for the block version. I am not sure if it would be that useful.

**Comment by johnjansen** [comment]
awesome @bcardiff i will put it up on my whiteboard ;-)

**Comment by asterite** [comment]
In the future (or soon) we could support unpacking in `if` and `while`. Ruby 2.3 or 2.4 supports this already. Basically:

```cr
if x, y = foo
end
```

is the same as this:

```cr
tmp = foo
if tmp
  x, y = tmp
end
```

So returning `nil` in partition would be good.

**Comment by johnjansen** [comment]
ok @asterite i'll get to it in a little bit ... also note the `partition` PR ;-)

**Comment by asterite** [comment]
@johnjansen In any case I don't know what final shape I'd like it to be. I'd like to know real use cases for partition before making a choice. What's your use case?

**Comment by johnjansen** [comment]
i dont have one right now, but someone was complaining about its absence, and i wanted to start contributing with something simple, so i decided to have a go

**Comment by johnjansen** [comment]
@asterite ... and now i do ... in the process of porting some code, i found this

```
    private def may_shift_colon?
      return false unless @text.includes?(":")
      partitions = @text.partition(":")
      partitions.last[0] !~ /\A\d+/ || partitions.first[-1] !~ /\A\d+/
    end
```

obviously, im happy to handle this any way thats necessary, but its one example ;-)

**Comment by asterite** [comment]
I found this about a use case for partition:

https://m0j0.wordpress.com/2007/07/11/partition-saves-my-bacon/

So it seems Python has this too. It's also a bit more efficient than `split` because it doesn't allocate an array. I think we can add it and keep the same behaviour as in Ruby and Python. `rpatition` also works the same in Ruby and Python: two empty strings followed by the original string.

I think the reason behind that behaviour is this. Imagine we expect to receive a string with a colon and we want to get what's before and after **the first** colon. So:

```cr
"foo:bar".partition(":") # => {"foo", ":", "bar"}
"foo:bar:baz".partition(":") # => ["foo", ":", "bar:baz"}
```

Now if there's no colon, we assume it's at the end of the string (because we are searching from the start). So:

```cr
# think "foo:".rpartition(":") # => {"foo", ":", ""}
"foo".partition(":") # => {"foo", "", ""}
```

An empty string is returned instead of the separator to distinguish it from the case where the colon was actually there.

Now, if we want to do the same but for **the last** colon:

```cr
"foo:bar".rpartition(":")     # => {"foo", ":", "bar"}
"foo:bar:baz".rpartition(":") # => {"foo:bar", ":", "baz"}
```

If there's no colon, since we are searching from the back of the string, we assume it's at the beginning of the string. So:

```cr
# think ":foo".rpatition(":") # => {"", ":", "foo"}
"foo".rpartition(":") # => {"", "", "foo"}
```

So now I think it makes a lot of sense to do it like that :-)

File: src/string.cr [hunk 10]
```diff
@@ -2200,6 +2200,64 @@ class String
...
+  # ```
+  # "hello".rpartition("l")  # => {"hel", "l", "o"}
+  # "hello".rpartition("x")  # => {"", "", "hello"}
+  # "hello".rpartition(/.l/) # => {"he", "ll", "o"}
+  # ```
+  def rpartition(search : (Char | String)) : Tuple(String, String, String)
+    pos = self.rindex(search)
+    search_size = search.is_a?(Char) ? 1 : search.size
+
+    pre, mid, post = {"", "", ""}
```
Comment (asterite) [inline src/string.cr]:
```cr
pre = mid = post = ""
```

**Comment by Sija** [comment]
@asterite It's also useful for splitting file paths, i.e. getting last path segment, that is the filename.

**Comment by asterite** [comment]
@johnjansen Could you rebase this PR against master?

**Comment by johnjansen** [comment]
something went sideways on me in the merge ... ill sort it out tomorrow

**Comment by johnjansen** [comment]
@asterite @bcardiff  i think we are good to merge ... ?

**Comment by luislavena** [comment]
@johnjansen I'm not sure, as there are still listed 19 commits for the changes.

https://github.com/crystal-lang/crystal/pull/3429/commits

I might recommend you look at the [Contributing](https://github.com/crystal-lang/crystal/blob/master/Contributing.md#maintain-clean-pull-requests) document for instructions to perform a clean rebase.

**Comment by johnjansen** [comment]
thanks @luislavena,  i hate rebase ;-) always messes with my head

**Comment by asterite** [comment]
@johnjansen Don't worry, I also hate it too :-)

I actually tried to manually rebase this PR but failed... or, mmm... I'm lazy to rewrite history.

Another alternative you can do is to start from scratch: close this PR and create a new one in a different branch.

**Comment by luislavena** [comment]
@johnjansen I use rebase all the time, but for some reason you're getting other commits that shouldn't be part of this branch.

I recommend taking a look to the document I linked before about updating your local copy against the upstream master and ensure **only** your commits are in the list of the rebase.

Perhaps use the interactive mode? (`git rebase -i upstream/master`)

**Comment by johnjansen** [comment]
@asterite im going to close this one, and open another, with just the stuff we want :-)
